bears/jinja2/Jinja2Bear.py

```diff
@@ -0,0 +1,271 @@
+            before=content_before,
+            spacing=spacing,
+            after=content_after,
+            content=match_object.group("content").strip(),
+            open=match_object.group("open"),
+            close=match_object.group("close")))
+    diff.change_line(
+        line_number,
+        line,
+        replacement)
+    return {filename: diff}
+
+
+def generate_label_diff(file, filename, line, line_number,
+                        match_object, expected_label):
+    '''
+    Generates a diff for a missing or wrong control loop end label.
+    Missing labels will be added, wrong ones replaced, content after the
+    label will be left untouched.
+
+
+    :param match_object:   the match object generated by a match on the regular
+                           expression control_end_regex in the run method
+    :param expected_label: The expected label for that control block.
+
+    '''
+    diff = Diff(file)
+
+    content_before = line[:match_object.end("close")]
+
+    suffix_start = match_object.end("close")
+    # if a label is there, we cut it out, by starting the kept suffix after it
+    if match_object.group("label") is not None:
+        suffix_start = match_object.end("label")
+
+    content_after = line[suffix_start:]
+
+    replacement = "{before}{label}{after}".format(
+        before=content_before,
+        after=content_after,
+        label=expected_label
+    )
+    diff.change_line(
+        line_number,
+        line,
+        replacement)
+    return {filename: diff}
+
+
+def is_spacing_valid(string, required_spacing):
+    '''
+    Returns True iff `string` has exactly <required_spacing>
+    spaces at the start and at the end
+    '''
+    leading_spaces = len(string) - len(string.lstrip())
+    trailing_spaces = len(string) - len(string.rstrip())
+
+    return (
+        leading_spaces == required_spacing
+        and trailing_spaces == required_spacing)
+
+
+class Jinja2Bear(LocalBear):
+    LANGUAGES = {"Jinja2"}
+    AUTHORS = {'The coala developers'}
+    AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
+    LICENSE = 'AGPL-3.0'
+    CAN_DETECT = {'Syntax'}
+    CAN_FIX = {'Formatting', 'Documentation'}
+
+    def handle_control_spacing_issue(self, file, filename, line, line_number,
+                                     control_spacing, match_object):
+        '''
+        Handles incorrectly spaced control tags (if and for),
+        both start and end tags
+
+        :param control_spacing: The number of spaces placed around control tags
+                                as required by the settings.
+        :param match_object:    the match object generated by a match on the
+                                regular expression control_start_regex or
+                                control_end_regex in the run method.
+        '''
+        diff = generate_spacing_diff(
+            file, filename, line, line_number, match_object, control_spacing)
+        return Result.from_values(
```

Comment:
btw. you can also use `self.new_result` and not have to pass `origin=self` as it's set implicitly  (it's a prop that binds the param to the returning closure)